frappe/public/js/frappe/form/share.js

```diff
@@ -112,9 +117,15 @@ frappe.ui.form.Share = Class.extend({
 				label: __("Share With"),
 				fieldname: "share_with",
 				options: "User",
-				filters: {
-					"user_type": "System User",
-					"name": ["!=", user]
+				get_query: function() {
+					return {
+						query: "frappe.core.doctype.docshare.docshare.docshare_user_query",
```

Comment:
Instead of re-implementing the user query, why not modify the user_query in frappe to accept filters and pass `name: ["not in", [list of existing shares that can be found on the client side itself]]`?

See:
1. https://github.com/frappe/frappe/blob/develop/frappe/core/doctype/user/user.py#L440
1. Move this function to frappe/...../reportview.py: https://github.com/frappe/erpnext/blob/develop/erpnext/controllers/queries.py#L10